src/Type/ArrayType.php

```diff
@@ -350,15 +350,19 @@ public static function castToArrayKeyType(Type $offsetType): Type
 				return new UnionType([new ConstantIntegerType(0), new ConstantIntegerType(1)]);
 			}
 
+			if ($offsetType instanceof UnionType) {
+				return $traverse($offsetType);
+			}
+
 			if ($offsetType instanceof FloatType || $offsetType->isNumericString()->yes()) {
 				return new IntegerType();
 			}
 
-			if ($offsetType instanceof StringType || $offsetType->isNonEmptyString()->yes()) {
+			if ($offsetType instanceof StringType || $offsetType->isString()->yes()) {
```

Comment:
StringType will also answer `isString()->yes()`, so I guess the BooleanOr isn't needed here?